### PR TITLE
Add support for periodically archiving the transaction billing table

### PR DESCRIPTION
**Pull request by hodgestar**
The transaction billing table has an entry per message which means it'll get large very quickly. We should archive data regularly (probably at the end of each month we should archive data older than N months).

### CODE REVIEW

**Comment by hodgestar** [comment]
This PR got a little big. It's ready for some reviewing but I'll split it up into separate pieces in the morning.

**Comment by hodgestar** [comment]
Depends on #1079 and #1080 but otherwise ready for review.

**Comment by hodgestar** [comment]
I still need to add gzip support, but I think that's best left for a separate PR.

**Comment by hodgestar** [comment]
#1079 and #1080 have landed. Gzip support is happening separately in #1085.

Ready for review.

File: go/billing/models.py [hunk 1]
```diff
@@ -351,3 +351,45 @@ class LineItem(models.Model):
...
+
+    STATUS_ARCHIVE_CREATED = 'archive_created'
+    STATUS_TRANSACTIONS_UPLOADED = 'transactions_uploaded'
+    STATUS_ARCHIVE_COMPLETED = 'archive_completed'
+    STATUS_ARCHIVE_DELETED = 'archive_deleted'
+    STATUS_CHOICES = tuple(
+        (choice, choice.title()) for choice in (
+            STATUS_ARCHIVE_CREATED, STATUS_TRANSACTIONS_UPLOADED,
+            STATUS_ARCHIVE_COMPLETED, STATUS_ARCHIVE_DELETED)
+    )
```
Comment (justinvdm) [inline go/billing/models.py:368]: Are we using `.title()` to turn these into readable labels? If we are, maybe we should be removing the `_`s.

File: go/billing/models.py [hunk 2]
```diff
@@ -351,3 +351,45 @@ class LineItem(models.Model):
...
+                    "(inclusive)."))
+
+    to_date = models.DateField(
+        help_text=_("The end of the date range covered by this archive "
+                    "(inclusive)"))
+
+    status = models.CharField(
+        max_length=32, choices=STATUS_CHOICES, default=STATUS_ARCHIVE_CREATED,
+        help_text=_("The status of this archive. One of pending, "
+                    "completed, or deleted."))
```
Comment (justinvdm) [inline go/billing/models.py]: Not seing 'pending' as one of the statuses, did this change? Or is that status implicit?

File: go/billing/tasks.py [hunk 3]
```diff
@@ -6,12 +6,33 @@
...
-    Account, Transaction, MessageCost, Statement, LineItem)
+    Account, MessageCost, Transaction, Statement, LineItem, TransactionArchive)
+from go.billing.django_utils import TransactionSerializer
 from go.base.utils import vumi_api
 
 
+def month_range(months_ago=1, today=None):
+    """ Return the dates at the start and end of a calendar month.
+
+    :param in months_ago:
```
Comment (justinvdm) [inline go/billing/tasks.py]: s/in/int

File: go/billing/tests/test_tasks.py [hunk 4]
```diff
@@ -238,3 +259,132 @@ def test_generate_monthly_statement_unaccessible_tags(self):
...
+            "transactions-test-0-user-2013-11-01-to-2013-11-30.json")
+        self.assertEqual(archive.from_date, from_date)
+        self.assertEqual(archive.to_date, to_date)
+        self.assertEqual(archive.status, archive.STATUS_ARCHIVE_COMPLETED)
+
+        self.assert_remaining_transactions(
+            transactions_before | transactions_after)
+        self.assert_archive_in_s3(
+            bucket, archive.filename,
+            sorted(transactions_within, key=lambda trans: trans.pk))
```
Comment (justinvdm) [inline go/billing/tests/test_tasks.py]: We don't seem to test that we set the statuses properly (from created to uploaded to completed). It doesn't seem like something easy to test though. Happy with that being untested if there's no non-contrived way to test that.

**Comment by justinvdm** [comment]
Left one more minor comment, but looks good otherwise.

File: go/billing/tests/test_tasks.py (hunk 4 above)
Comment (hodgestar) [inline go/billing/tests/test_tasks.py]: I realised that I can solve this by adding a `delete` parameter to archiving which I was adding anyway in the management command branch.

**Comment by hodgestar** [comment]
Minor comment addressed, I think.

**Comment by justinvdm** [comment]
:+1:

**Comment by justinvdm** [comment]
err, or rather, :+1: once travis is happy.

**Pull request merged**